Adding ECDSA support

diogomonica opened the pull request:
Added support for ECDSA keys.

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "adding-ecdsa" git@github.com:docker/notary.git somewhere
$ cd somewhere
$ git rebase -i HEAD~4
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

aaronlehmann commented on client/cli_crypto_service.go:
```diff
@@ -10,62 +11,77 @@ import (
-		// Read PrivateKey from file
+		// Read PrivateKey from file.
+		// TODO(diogo): This assumes both that only root keys are encrypted and
+		// that encrypted keys are always X509 encoded
```
I do think we'll want to change this eventually - I can imagine use cases where the root keys will not have a passphrase (i.e. non-interactive use). I think it would be a good idea to add a flag to the `genericCryptoService` structure to use instead of using the passphrase for deciding whether it's a root key or not.

diogomonica replied:
Agreed.

aaronlehmann commented on client/cli_crypto_service.go:
```diff
@@ -95,9 +120,8 @@ func (ccs *CryptoService) Sign(keyIDs []string, payload []byte) ([]data.Signatur
-	// TODO(diogo): Implement support for ECDSA.
-	if strings.ToLower(privKey.Cipher()) != "rsa" {
+func rsaSign(privKey *data.PrivateKey, hash crypto.Hash, hashed []byte) ([]byte, error) {
+	if strings.ToUpper(privKey.Cipher()) != "RSA" {
```
I wonder if it would make sense to have `Cipher` do the `ToUpper` so we don't have the risk of forgetting to transform to uppercase for a string comparison.

I also think it would be a good idea to make the algorithms string constants we can refer to (something like `signer.AlgoRSA`) to guard against incorrect character case or typos, but this is more of a cosmetic thing. It would also help prevent confusion between cryptographic algorithms and signing methods.

endophage replied:
+1 for the string constants.

Having Cipher do the case conversion seems like it would be forcing a convention on everyone. Whatever somebody sets as the type when creating a key should be what they get back when calling Cipher.

aaronlehmann commented on client/client_test.go:
```diff
@@ -18,6 +18,8 @@ import (
 )
 
 const timestampKeyJSON = `{"keytype":"RSA","keyval":{"public":"MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAyyvBtTg2xzYS+MTTIBqSpI4V78tt8Yzqi7Jki/Z6NqjiDvcnbgcTqNR2t6B2W5NjGdp/hSaT2jyHM+kdmEGaPxg/zIuHbL3NIp4e0qwovWiEgACPIaELdn8O/kt5swsSKl1KMvLCH1sM86qMibNMAZ/hXOwd90TcHXCgZ91wHEAmsdjDC3dB0TT+FBgOac8RM01Y196QrZoOaDMTWh0EQfw7YbXAElhFVDFxBzDdYWbcIHSIogXQmq0CP+zaL/1WgcZZIClt2M6WCaxxF1S34wNn45gCvVZiZQ/i ...
```
Looks like we aren't using `timestampKeyJSON` anymore. I guess this is because the timestamp key is hardcoded to ECDSA right now. Could you please add a TODO comment saying to re-add test coverage with the RSA timestamp key once it's possible to use RSA for a timestamp key?

diogomonica replied:
I tried. After 1 hour of debugging turns out that this is happening:

- We create two signers, one for the root key and one for other keys
- We indiscriminately call both signers with all keys and what works works (if signatures get appended great, if not, tough luck)
- We are able to handle the mode of Root RSA and targets/snapshots ECDSA, and root ECDSA and targets/snapshots ECDSA because they have different signers and are always aligned on the key type.
- Now the timestamps key comes in. Since the timestamp key is generated remotely, it has to be aligned with whatever the snapshot/targets are using. This means that if we have an RSA timestamp key, and generate ECDSA targets/snapshots, we will always fail because the cryptoservice that has access to that key is an ECDSA cryptoservice. The opposite is also true.

Currently in the code we're hardcoding ECDSA targets and snapshot keys.

aaronlehmann commented on trustmanager/x509utils.go:
```diff
@@ -125,8 +95,19 @@ func fingerprintCert(cert *x509.Certificate) CertID {
+	case x509.ECDSA:
+		keyType = "ECDSA"
+	default:
+		logrus.Debug("error while fingerprinting certificate. Got Unknown key type.")
```
Let's return an explicit error here.

diogomonica replied:
I had to rework a ton of things that expected this method to never error out, but done.

GordonTheTurtle commented:
Looks like we would not be able to merge this PR because of merge conflicts. Please rebase, fix conflicts, and force push to your branch.

diogomonica commented:
@endophage @aaronlehmann addressed all the comments except refactoring the passphrase checking. We can do that tomorrow.

Also, need review of the remainer of https://github.com/endophage/gotuf/pull/9 to then be able to `godep save` on this branch and merge to master.
